sktime/transformers/series_as_features/signature_based/_window.py

```diff
@@ -0,0 +1,189 @@
+
+    Returns
+    -------
+    list:
+        A list of lists where the inner lists are lists of tuples that
+        denote the start and end indexes of each window.
+    """
+    # Setup all available windows here
+    length_step = {'length': window_length, 'step': window_step}
+    window_dict = {
+        'global': (Global, {}),
+        'sliding': (Sliding, length_step),
+        'expanding': (Expanding, length_step),
+        'dyadic': (Dyadic, {'depth': window_depth}),
+    }
+
+    assert window_name in window_dict.keys(), (
+        "Window name must be one of: {}. Got: {}."
+        "".format(window_dict.keys(), window_name)
+    )
+
+    window_cls, window_kwargs = window_dict[window_name]
+    return window_cls(**window_kwargs)
+
+
+class Window:
+    """Abstract base class for windows.
+
+    Each subclass must implement a __call__ method that returns a list of lists
+    of 2-tuples. Each 2-tuple specifies the start and end of each window.
+    These windows are grouped into a list that will (usually) cover the full
+    time series. These lists are grouped into another list for situations
+    where we consider windows of multiple scales.
+    """
+    def num_windows(self, length):
+        """Method that returns the total number of windows in the set.
+
+        Parameters
+        ----------
+        length: int, The length of the input path.
+
+        Returns
+        -------
+        int: The number of windows.
+        """
+        return sum([len(w) for w in self(length)])
+
+
+class Global(Window):
+    """ A single window over the full data. """
+    def __call__(self, length=None):
+        return [[_Pair(None, None)]]
+
+
+class _ExpandingSliding(Window):
+    def __init__(self, initial_length, start_step, end_step):
+        """
+        Parameters
+        ----------
+        initial_length: int, Initial length of the input window.
+        start_step: int, Initial step size.
+        end_step: int, Final step size.
+        """
+        super(_ExpandingSliding, self).__init__()
+        self.initial_length = initial_length
+        self.start_step = start_step
+        self.end_step = end_step
+
+    def __call__(self, length):
+        def _call():
+            start = 0
+            end = self.initial_length
+            while end <= length:
+                yield _Pair(start, end)
+                start += self.start_step
+                end += self.end_step
+        return [list(_call())]
+
+
+class Sliding(_ExpandingSliding):
+    """A window starting at zero and going to some point that increases
+    between windows.
+    """
+    def __init__(self, length, step):
+        """
+        Parameters
+        ----------
+        length: int, The length of the window.
+        step: int, The sliding step size.
+        """
+        super(Sliding, self).__init__(
+            initial_length=length, start_step=step, end_step=step
+        )
+
+
+class Expanding(_ExpandingSliding):
```

Comment:
same here, I'd suggest making this a private class, unless it's meant to be used directly by users, otherwise perhaps a somewhat more descriptive name